storage/rocksdb/ha_rocksdb.cc

```diff
@@ -8443,7 +8482,9 @@ int ha_rocksdb::read_row_from_secondary_key(uchar *const buf,
       if (pk_size == RDB_INVALID_KEY_LEN) {
         rc = HA_ERR_ROCKSDB_CORRUPT_DATA;
       } else {
-        rc = get_row_by_rowid(buf, m_pk_packed_tuple, pk_size);
+        if (!covered_lookup || m_lock_rows != RDB_LOCK_NONE) {
```

Comment:
@Pushapgl We're trying to resolve this diff with covered index lookups. If we enter this code path, and covered_lookup is true, then it seems like we should also call global_stats.covered_secondary_key_lookups.inc();

This doesn't appear to be the case in 5.6 either. Is this a bug?

Reply:
Actually, find_icp_matching_index_rec() advances the key until an ICP match is found. Doesn't this mean covered_lookup needs to be recalculated based on the new key?

Reply:
cc: @spetrunia

Reply:
> Actually, find_icp_matching_index_rec() advances the key until an ICP match is found. Doesn't this mean covered_lookup needs to be recalculated based on the new key?

Yes, correct. The current code has a bug.
if `have_icp==true`, and we start from a row for which `covered_lookup=true`, and go into `find_icp_matching_index_rec`, which will move to the row which should have `covered_lookup=false`  and return, then we will fail to call get_row_by_rowid(). 

(I'm surprised that ICP is used for such partially-covered keys, though.. I that it wasn't)

Reply:
... welll apparently it is used. (and there's nothing wrong with that.. there might be difficulties if the range we are scanning uses the partially-covered column, but it's not the case her)

Reply:
In the case where `covered_lookup == false` and `have_icp == false`, it seems entering the `else` clause and calling find_icp_matching_index_rec() would be the wrong action. Reading the full row is still needed though.

It seems this function should be re-ordered to account for all of these conditions correctly. Something like:

```
int ha_rocksdb::read_row_from_secondary_key(uchar *const buf,
                                            const Rdb_key_def &kd,
                                            bool move_forward) {
  int rc = 0;
  uint pk_size;

  // Due to MRR, now an index-only scan have pushed index condition.
  // (If it does, we follow non-index only code path here, except that
  //  we don't fetch the row).
  bool have_icp = (pushed_idx_cond && pushed_idx_cond_keyno == active_index);
  if (have_icp) {
    if (kd.m_is_reverse_cf) move_forward = !move_forward;
    rc = find_icp_matching_index_rec(move_forward, buf);
    if (rc) return (rc);
  }

  /* Get the key columns and primary key value */
  const rocksdb::Slice &rkey = m_scan_it->key();
  const rocksdb::Slice &value = m_scan_it->value();

#ifndef DBUG_OFF
  bool save_keyread_only = m_keyread_only;
#endif
  DBUG_EXECUTE_IF("dbug.rocksdb.HA_EXTRA_KEYREAD", { m_keyread_only = true; });

  bool covered_lookup = (m_keyread_only && kd.can_cover_lookup()) ||
                         kd.covers_lookup(&value, &m_lookup_bitmap);

#ifndef DBUG_OFF
  m_keyread_only = save_keyread_only;
#endif

  pk_size =
    kd.get_primary_key_tuple(table, *m_pk_descr, &rkey, m_pk_packed_tuple);
  if (pk_size == RDB_INVALID_KEY_LEN) {
    rc = HA_ERR_ROCKSDB_CORRUPT_DATA;
  } else {
    if (covered_lookup && m_lock_rows == RDB_LOCK_NONE) {
      rc = kd.unpack_record(table, buf, &rkey, &value,
                            m_converter->get_verify_row_debug_checksums());
      global_stats.covered_secondary_key_lookups.inc();
    } else {
      rc = get_row_by_rowid(buf, m_pk_packed_tuple, pk_size);
    }
  }

  if (!rc) {
    m_last_rowkey.copy((const char *)m_pk_packed_tuple, pk_size,
                       &my_charset_bin);
  }

  return rc;
}
```

@spetrunia , For the MRR case, the comment is that if `have_icp == true`, then we want to skip fetching the row. But the original code doesn't check for have_icp before fetching the row. If `have_icp == true`, can we just call kd.unpack_record() and change the conditional from `if (covered_lookup && m_lock_rows == RDB_LOCK_NONE)` to `if ((covered_lookup || have_icp) && m_lock_rows == RDB_LOCK_NONE)`.

Reply:
I just tested this. Skipping `get_row_by_rowid()` based on `have_icp == true` will return the wrong data.